cmd/dockerd/tracing.go

```diff
@@ -0,0 +1,156 @@
+package main
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"sync"
+
+	"github.com/containerd/containerd/log"
+	"github.com/moby/buildkit/util/tracing/detect"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+// See https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/ for details on env vars/values.
+const (
+	otelSDKDisabledEnv                = "OTEL_SDK_DISABLED"
+	otelTracesExporterEnv             = "OTEL_TRACES_EXPORTER"
+	otelExporterOTLPEndpointEnv       = "OTEL_EXPORTER_OTLP_ENDPOINT"
+	otelExporterOTLPTracesEndpointEnv = "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT"
+	otelExporterOTLPTracesProtocol    = "OTEL_EXPORTER_OTLP_TRACES_PROTOCOL"
+	otelExporterOTLPProtocolEnv       = "OTEL_EXPORTER_OTLP_PROTOCOL"
+	otelServiceNameEnv                = "OTEL_SERVICE_NAME"
+	otelTracesSamplerEnv              = "OTEL_TRACES_SAMPLER"
+	otelTracesSamplerArgEnv           = "OTEL_TRACES_SAMPLER_ARG"
+)
+
+var errTracingDisabled = errors.New("tracing disabled")
+
+func getTracerProvider(ctx context.Context, getEnv func(string) string) (*sdktrace.TracerProvider, error) {
+	// By default the OTLP libs will connect to localhost if no endpoint is set.
```

Comment:
After much head->desk:

Turns out, all the difference is this code was defaulting to http/prorobuf for OTLP, detect defaults to grpc.
This is why I wasn't getting traces in my exporter.
Per spec, default should be http/protobuf: https://opentelemetry.io/docs/specs/otel/protocol/exporter/#specify-protocol

I don't want to break existing users of buildkit traces... but also it seems more correct to go with the spec for tracing the daemon.... anyone have any thoughts on this?

Reply:
Spec:

> If no configuration is provided the default transport SHOULD be http/protobuf unless SDKs have good reasons to choose grpc as the default (e.g. for backward compatibility reasons when grpc was already the default in a stable SDK release).

Reply:
I don't remember where I took this code anymore. I thought it might have something to do with UNIX and HTTP not working well together for the socket we send to containers, but that code sets a specific protocol. https://github.com/moby/buildkit/blob/e2dfabdebce24c9f55623259160a6a9faa1cb31e/executor/oci/spec.go#L44

Assuming the only thing it can break is external OTLP collectors that don't follow spec, I think we should just change it in buildkit. However, it wouldn't be nice to change it in a patch release.

Reply:
When I wrote the `detect` package, the (implicit) default was still grpc.
https://github.com/open-telemetry/opentelemetry-specification/pull/1969/commits/0297525a5634e233533bc5b14b15b69fae18abb8
https://github.com/moby/buildkit/commit/292f1faf1209bb2a49fa4a3e9c3f785c51afbe55#diff-77fb71b07405102c7882eae6d80aa8300966cb29ffe2d478a63b46899602f1f8R29

Reply:
Yep, definitely not surprised the default changed 😸 

I've updated this PR to use detect for everything and also set the otlp protocol to http/profobuf if unset so that `detect` will pick it up.
LMK if this sounds good.